Port the MainWindow ui to a .ui file

leia-uwu opened the pull request:
Some stuff still needs to be done in the c++ side because qt designer is dumb
(The instance toolbar icon and name buttons, and some shortcuts are still manually added in MainWindow.cpp)

Looks almost identical, with some minor tweaks:
- The instance toolbar has been changed to a WideBar, allowing for customization of actions.
- The instance toolbar buttons are now fullwidth.
- The close window action has been moved to the end of the file menu and now has an icon.
- The help menu has some layout changes.

This also fixes some stuff:
- Menus not having tooltips.
- The top toolbar not connecting to the title bar in kde.
- The instance toolbar separators looking weird after you move the toolbar.

Closes #594, closes #69, closes #473

leia-uwu committed:
feat: add Widebar::InsertWidgetBefore method

Signed-off-by: leo78913 <leo3758@riseup.net>

leia-uwu committed:
refactor: Port the main window to a .ui file

some stuff still needs to be done in the c++ side because qt designer is dumb >:(

the instance toolbar icon and instance name buttons are still added manually inside MainWindow.cpp

looks almost identical, with some minor tweaks:
- the instance toolbar is now a WideBar, so you can customize what actions you want :D
- the instance toolbar buttons are now fullwidth
- the close window button is now at the end of the file menu
- the help menu has some layout changes

this also fixes some stuff:
- menus not having tooltips
- the top toolbar not connecting to the title bar in kde
- the instance toolbar separators looking weird after you move the toolbar

Signed-off-by: leo78913 <leo3758@riseup.net>

leia-uwu committed:
feat(WideBar): Allow disabling alt shortcuts

Signed-off-by: leo78913 <leo3758@riseup.net>

leia-uwu committed:
fix(widebar): fix insertSeparator

WideBar::insertSeparator was adding the separator to the end of the toolbar

Signed-off-by: leo78913 <leo3758@riseup.net>

flowln commented on launcher/ui/MainWindow.cpp:
```diff
@@ -1375,14 +698,10 @@ void MainWindow::repopulateAccountsMenu()
-        ui->actionNoAccountsAdded = TranslatedAction(this);
-        ui->actionNoAccountsAdded->setObjectName(QStringLiteral("actionNoAccountsAdded"));
-        ui->actionNoAccountsAdded.setTextId(QT_TRANSLATE_NOOP("MainWindow", "No accounts added!"));
+        ui->actionNoAccountsAdded->setText( "No accounts added!");
```
```suggestion
        ui->actionNoAccountsAdded->setText(tr("No accounts added!"));
```

flowln commented on launcher/ui/MainWindow.cpp:
```diff
@@ -1414,18 +733,17 @@ void MainWindow::repopulateAccountsMenu()
+    ui->actionNoDefaultAccount = new QAction(this);
     ui->actionNoDefaultAccount->setObjectName(QStringLiteral("actionNoDefaultAccount"));
-    ui->actionNoDefaultAccount.setTextId(QT_TRANSLATE_NOOP("MainWindow", "No Default Account"));
+    ui->actionNoDefaultAccount->setText("No Default Account");
```
```suggestion
    ui->actionNoDefaultAccount->setText(tr("No Default Account"));
```

flowln commented on launcher/ui/MainWindow.cpp:
```diff
@@ -139,785 +140,100 @@ QString profileInUseFilter(const QString & profile, bool used)
+        // disabled until we have an instance selected
+        ui->instanceToolBar->setEnabled(false);
+        ui->actionKillInstance->setEnabled(false);
+        ui->actionLaunchInstance->setEnabled(false);
```
those two don't seem to have any impact, removing them still leaves the actions disabled when no instance is selected :p

flowln commented on launcher/ui/MainWindow.cpp:
```diff
@@ -929,12 +244,13 @@ MainWindow::MainWindow(QWidget *parent) : QMainWindow(parent), ui(new MainWindow
-        ui->newsToolBar->insertWidget(ui->actionMoreNews, newsLabel);
+        newsLabel = dynamic_cast<QToolButton*>(ui->newsToolBar->widgetForAction(ui->actionNewsLabel));
+
+        //add a spacer before the more news button
```
i think the previous behavior was a bit better, you had a lot more room to click the news entry if you wanted, space that otherwise is just empty.

leia-uwu replied:
idk how to fix that without adding the news label toolbutton manually, i tried setting the size policy but it has no effect :/

flowln replied:
seems like it indeed will not be possible without manually inserting the toolbutton, since it gets added via `addAction` in the generated file, which i think forces the layout to be that way. Still, I think it is still worth it to manually add it :slightly_smiling_face:

flowln reviewed (commented):
this looks pretty good overall!

there's only a couple minor things that I think could be improved, outside the comments i've made:
1. While you did fix the alt shortcuts on the WideBar, they still show up in the Right-click menu, even though you can't even use those without closing the menu :p
![image](https://user-images.githubusercontent.com/9145768/211418673-69b6b8aa-b342-471a-bc59-f74d507c6253.png)

2. I'm not sure about having icons in that menu. In the instance WideBars, this was not an issues because the actions don't have icons, but here they do, and it becomes ambiguous whether clicking that option will toggle its visibility, or activate the action.

3. Maybe the old right-click menu actions, now only accessible through the top bar and the news bar, should be merged with this new menu somehow. It is quite weird to not be able to hide the Instance toolbar from interacting with that toolbar, but be able to by interacting with other toolbars. Perhaps they could be separated into two different sections, and the one to toggle the actions visibility have a name that would clear up the ambiguity I talked about in item 2, so the icons could be left there :)

leia-uwu committed:
Fix: translate NoAccountsAdded text

Co-authored-by: flow <flowlnlnln@gmail.com>
Signed-off-by: leo78913 <leo3758@riseup.net>

leia-uwu committed:
Fix: translate actionNoDefaultAcount text

Co-authored-by: flow <flowlnlnln@gmail.com>
Signed-off-by: leo78913 <leo3758@riseup.net>

leia-uwu commented:
> 1. even though you can't even use those without closing the menu :p

uh, for some reason the menu only closes when you press alt while using the fusion theme in my testing, this also happens in the instance view context menu and in the menubar :/

leia-uwu committed:
feat(WideBar): custom context menu actions

Signed-off-by: leo78913 <leo3758@riseup.net>

leia-uwu committed:
remove useless setEnabled calls

Signed-off-by: leo78913 <leo3758@riseup.net>

leia-uwu commented:
:catstare:
![image](https://user-images.githubusercontent.com/69596237/211587730-2e6d1f58-792a-462c-8b82-4f2570ec7770.png)

DioEgizio commented:
with this the status bar doesn't seem to follow my prism theme

<img width="1143" alt="this pr" src="https://user-images.githubusercontent.com/83089242/211620850-aa120d15-f1e0-4e38-a273-d438cdd78685.png"> this pr

<img width="1606" alt="w/o this pr" src="https://user-images.githubusercontent.com/83089242/211620909-96c97e85-2955-42c3-abd4-02cd0ba425ae.png"> without this pr

TayouVR commented:
hmmm... its a different shade of grey and doesn't have the margins/paddings from the theme. 
After scrolling through the code I can't spot any obvious reasons for that, maybe the various size constraints on the QStatusBar in the .ui are different from the defaults it had in code? 🤔 
![image](https://user-images.githubusercontent.com/31988415/211622876-0fcffdf3-d3d3-4512-9a33-67b0ddfe45c7.png)

leia-uwu committed:
Fix status bar name

Signed-off-by: leo78913 <leo3758@riseup.net>

leia-uwu commented:
> with this the status bar doesn't seem to follow my prism theme

Fixed

Scrumplex reviewed (approved):
I couldn't find any differences, besides the news toolbar as @flowln already pointed out.
LGTM and thanks!

leia-uwu committed:
Fix: make the newsLabel toolbutton fullwidth again

this reverts it to how it was before the MainWindow .ui port

Signed-off-by: leo78913 <leo3758@riseup.net>

TayouVR reviewed (approved):
couldn't find any issues in my testing, looking good!

kumquat-ir commented on launcher/ui/MainWindow.cpp:
```diff
@@ -139,785 +140,101 @@ QString profileInUseFilter(const QString & profile, bool used)
-            MainWindow->setObjectName(QStringLiteral("MainWindow"));
-        }
-        MainWindow->resize(800, 600);
-        MainWindow->setWindowIcon(APPLICATION->getThemedIcon("logo"));
```
This seems to have been removed without a replacement, the window icon is missing (left is this pr, right is `aur/prismlauncher-git`)
![Screenshot_2023-01-17_02-27-38](https://user-images.githubusercontent.com/66188216/212835620-53a5e38c-c6ed-4195-b245-e0aada6b4e23.png)

leia-uwu committed:
Fix main window icon and stuff

i forgor 💀

Signed-off-by: leo78913 <leo3758@riseup.net>

leia-uwu committed:
refactor: clean up some MainWindow stuff

this makes the accounts button and menubar item share the same QMenu
and also refactors some code

Signed-off-by: leo78913 <leo3758@riseup.net>

flowln commented on launcher/ui/widgets/WideBar.cpp:
```diff
@@ -133,7 +158,9 @@ void WideBar::insertSeparator(QAction* before)
 
     BarEntry entry;
-    entry.bar_action = QToolBar::insertSeparator(before);
+    entry.bar_action = new QAction("", this);
+    entry.bar_action->setSeparator(true);
+    insertAction(iter->bar_action, entry.bar_action);
```
why is this change needed?

leia-uwu replied:
it was adding the separator to the end of the toolbar instead of before the action

flowln replied:
that's because the old call was using `before` as the action to insert in relation to, i believe. It should use `iter->bar_action` instead, since that's the actual action in the WideBar.

leia-uwu replied:
done

flowln commented on launcher/ui/widgets/WideBar.cpp:
```diff
@@ -111,6 +126,16 @@ void WideBar::insertActionAfter(QAction* after, QAction* action)
+    if (iter == m_entries.end())
+        return;
+
+    BarEntry entry;
+    entry.bar_action = insertWidget(iter->bar_action, widget);
```
what's the use of `entry` here?

leia-uwu replied:
oops

leia-uwu replied:
fixed

flowln commented on launcher/ui/MainWindow.ui:
```diff
@@ -0,0 +1,696 @@
+    <string>&amp;Kill</string>
+   </property>
+   <property name="toolTip">
+    <string>Kill the running instance</string>
```
```suggestion
    <string>Kill the running instance.</string>
```

flowln commented on launcher/ui/MainWindow.ui:
```diff
@@ -0,0 +1,696 @@
+    <string>&amp;Matrix Space</string>
+   </property>
+   <property name="toolTip">
+    <string>Open %1 Matrix space</string>
```
```suggestion
    <string>Open %1 Matrix space.</string>
```

flowln reviewed (commented):
some nitpicks, but looks good overall!

edgars-cirulis commented:
Kind of annoying, after toggling, the check automatically closes the menu.

leia-uwu committed:
Update launcher/ui/MainWindow.ui

Co-authored-by: flow <flowlnlnln@gmail.com>
Signed-off-by: leo78913 <leo3758@riseup.net>

leia-uwu committed:
Update launcher/ui/MainWindow.ui

Co-authored-by: flow <flowlnlnln@gmail.com>
Signed-off-by: leo78913 <leo3758@riseup.net>

leia-uwu committed:
Remove unused BarEntry variable

Signed-off-by: leo78913 <leo3758@riseup.net>

leia-uwu committed:
better fix for WideBar::insertSeparator

Signed-off-by: leo78913 <leo3758@riseup.net>

flowln reviewed (approved):
lgtm, thanks!
